### ISSUE TITLE
Ability to disable hash suffix for generated secrets and configmaps.

### ISSUE THREAD

**Issue by mattatcha** [kubernetes-sigs/kustomize#50]
Requiring a rolling update due to config changes isn't always an option. 

For example most prometheus deployments are a single instance and recreating the pod for every config change would cause "holes" in time series metrics.

**Comment by schweikert** [comment]
It seems to me that while having the hash suffix is often a good thing and something that you want,  you might have situations where you don't want that feature. Maybe it could be made optional in some way, maybe with an annotation or something like that.

**Comment by pwhittlesea** [comment]
So I'm currently looking into this on my fork. My current solution is to add a renaming 'behavior' to the generated configmaps and generated secrets:

```yaml
configMapGenerator:
- name: ldap-configmap
  renaming: none
  files:
    - env.startup.txt
```

This approach works for config-maps and secrets that have been created using the generators but does not disable the renaming of ConfigMaps and Secrets defined via K8S Yaml.

I am thinking that a global flag to change the default renaming behavior to 'none' and then users which want to leverage the hash-suffix renaming could specify:

```yaml
configMapGenerator:
- name: ldap-configmap
  renaming: hash_suffix
  files:
    - env.startup.txt
```

@mattatcha Does this align with your original thoughts?

@Liujingfang1 Are you able to have a quick look at [my initial commit](https://github.com/pwhittlesea/kustomize/commit/849b27d95a923b2a7d261a67604e91d2ce987ca7)? I would appreciate feedback on the approach prior to me writing tests and documentation?

**Comment by pwhittlesea** [comment]
@Liujingfang1 I agree that a global flag inside the kustomize file would be preferable. Are you thinking something at the same level as the `namePrefix` option?

I would also like to put forward my reason for needing this feature:
I do use the rolling update for most of my configmaps, however my company provides our platform to customers with the ability for them to configure certain parts of the system. Due to certain regulatory requirements we have to ensure that the configuration is precisely documented and, as we use configmaps to provide this configuration, configmaps with random names are not suitable for us. Disabling this for some config maps would allow me to transition to kustomize without having to address the much larger issue.

I feel that, although having the rolling updates for free when using kustomize is a great thing, it is functionallity that maybe not all users want.

However I also accept that I am probably in a fringe set of users and kustomize (rightfully so) aims to be an opinionated tool that forces best practice on deployment.

**Comment by mxey** [comment]
> For the case using configmap/secret outside kustomization, is it possible to manage them inside the same kustomization?

No, the secret is used by various application deployments from different repositories.

**Comment by Liujingfang1** [comment]
@mxey @pwhittlesea Which way do you use more often for configmaps, from a configmap.YAML or from a `configmapGenerator`? What if we don't append the hash for a configmap read from a YAML file?

**Comment by pwhittlesea** [comment]
@Liujingfang1 That would work for me. Maybe a flag in the kustomize file defaulting to the current behaviour? This would effectively allow my team to treat config maps and secrets defined in YAML as 'constant' vs the managed ones in the generator sections that provide rolling update support.

**Comment by mxey** [comment]
@Liujingfang1 that would work for me as well.

**Comment by Liujingfang1** [comment]
This is the PR #159 for only adding hash for configmap/secretes created from generators

**Comment by pnovotnak** [comment]
I'd still like to see this. In my case, I have some objects in a separate repo that reference well known secrets. Since we keep kustomize configurations in a repository, we will either end up managing these secrets separately (eg: `{ kustomize ... ; decrypt ... ; } | kubectl apply -f-`) or having to run a script in the repository that we keep the configurations in, in order to have yamls on disk.

Regardless of specific use cases, IMO the current behavior is surprising and opposite what it should be. When generating a secret, `name` should equal what you see when you do a `kubectl get secrets`. A new field, `generateName`, should be what you use if you want to generate a name with a hash appended. And `generateName` should not involve any dashes magic... It should just append some version hash. This would make kustomize more consistent with the [current kubernetes api](https://kubernetes.io/docs/reference/federation/v1/definitions/#_v1_objectmeta)

**Comment by schweikert** [comment]
For the records: see also #489 , which is about implementing a global options for generators (including disabling the hash suffix)

**Comment by preetak** [comment]
Just went through the messages quickly. Do we have anything  for disabling the  appending of hash to the secret name generated ?

**Comment by natasha41575** [comment]
@preetak Yes, for example:

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
generatorOptions:
  disableNameSuffixHash: true
  labels:
    foo: bar
secretGenerator:
- name: shouldNotHaveHash
  literals:
  - foo=bar
```

It is exactly the same syntax for configMapGenerator.

**Comment by pnorth1** [comment]
Is it possible to disable the suffix hash for some, but not all, of the generated configmaps/secrets?

**Comment by pospospos2007** [comment]
> apiVersion: kustomize.config.k8s.io/v1beta1
> kind: Kustomization
> generatorOptions:
>   disableNameSuffixHash: true
>   labels:
>     foo: bar

You saved me. Love you!

**Comment by Dhruv-Soni** [comment]
> ~Is it possible to disable the suffix hash for some, but not all, of the generated configmaps/secrets?~
> 
> It is possible to disable the suffix hash on an individual configmap/secret basis with something like the following:
> 
> ```
> configMapGenerator:
>   - name: some-cmap
>     files:
>       - ./blah.yaml
>     options:
>       disableNameSuffixHash: true
> ```

Same, thanks, best code ever!

**Comment by agus7fauzi** [comment]
> Same, thanks, best code ever!

Same, thanks, best code ever! :))

**Comment by agus7fauzi** [comment]
> ~Is it possible to disable the suffix hash for some, but not all, of the generated configmaps/secrets?~
> 
> It is possible to disable the suffix hash on an individual configmap/secret basis with something like the following:
> 
> ```
> configMapGenerator:
>   - name: some-cmap
>     files:
>       - ./blah.yaml
>     options:
>       disableNameSuffixHash: true
> ```

Nice solution :)
